clients/adapter-kafka/src/main/java/org/eclipse/hono/adapter/client/telemetry/kafka/AbstractKafkaBasedMessageSender.java

```diff
@@ -0,0 +1,304 @@
+     */
+    protected final Logger log = LoggerFactory.getLogger(getClass());
+    protected final Tracer tracer;
+
+    private final Map<String, String> config;
+    private final KafkaProducerFactory<String, Buffer> producerFactory;
+    private final String producerName;
+
+    /**
+     * Creates a new Kafka-based message sender.
+     *
+     * @param producerFactory The factory to use for creating Kafka producers.
+     * @param producerName The producer name to use.
+     * @param config The Kafka producer configuration properties to use.
+     * @param tracer The OpenTracing tracer.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public AbstractKafkaBasedMessageSender(final KafkaProducerFactory<String, Buffer> producerFactory,
+            final String producerName,
+            final Map<String, String> config,
+            final Tracer tracer) {
+        Objects.requireNonNull(producerFactory);
+        Objects.requireNonNull(producerName);
+        Objects.requireNonNull(config);
+        Objects.requireNonNull(tracer);
+
+        this.producerFactory = producerFactory;
+        this.producerName = producerName;
+        this.config = config;
+        this.tracer = tracer;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Starts the producer.
+     */
+    @Override
+    public Future<Void> start() {
+        configureUniqueClientId();
+        getOrCreateProducer();
+        return Future.succeededFuture();
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Closes the producer.
+     */
+    @Override
+    public Future<Void> stop() {
+        return producerFactory.closeProducer(producerName);
+    }
+
+    /**
+     * Sends a message to a kafka cluster.
+     *
+     * @param topic The topic to send the message to.
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The device identifier.
+     * @param qos The delivery semantics to use for sending the data.
+     * @param payload The data to send.
+     * @param properties Additional meta data that should be included in the downstream message.
+     * @param context The currently active OpenTracing span (may be {@code null}). An implementation should use this as
+     *            the parent for any span it creates for tracing the execution of this operation.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future will be succeeded if the message has been sent downstream.
+     *         <p>
+     *         The future will be failed with a {@link org.eclipse.hono.client.ServerErrorException} if the data could
+     *         not be sent. The error code contained in the exception indicates the cause of the failure.
+     * @throws NullPointerException if topic, tenantId, deviceId or qos are {@code null}.
+     */
+    public Future<Void> send(final String topic, final String tenantId, final String deviceId,
```

Comment:
1. I mainly focused on changes required from the perspective of the command router. I like this idea of having two methods namely `send()` and `sendAndWaitForOutcome()` rather than passing the QoS and handling internally inside a single send method. With these new methods, the intentions are made clear.

2. In case of command router, the headers in the `KafkaConsumerRecord` can be reused while forwarding messages to the internal command topic. By having methods that accept headers, it avoids unnecessary conversion to properties from headers and then convert again to headers. FMPOV its good to have the above new methods overloaded also to accept headers.

Reply:
With the latest commit, I have introduced two types of methods namely `send()` and `sendAndWaitForOutcome()`.